spec/rspec/rails/matchers/have_http_status_spec.rb

```diff
@@ -177,6 +177,51 @@ def create_response(opts = {})
     end
   end
 
+  shared_examples_for 'status code matcher' do
+    # within the calling block, define:
+    # let(:expected_code) { <value> } # such as 555, 444, 333, 222
+    # let(:other_code) { <value> } # such as 555, 444, 333, 222
+    # let(:failure_message) {
+    #   /an error status code \(5xx\) but it was 400/
+    # } # for example
+    # let(:negated_message) {
+    #   /not to have an error status code \(5xx\) but it was 555/
+    # } # for example
+    # let(:description) {
+    #   'respond with an error status code (5xx)'
+    # } # for example
+    # subject(:matcher) { <matcher> } # e.g. { have_http_status(:error) }
+    describe "matching a response" do
+      it "returns true for a response with code" do
+        response = create_response(:status => expected_code)
+        expect(matcher.matches?(response)).to be(true)
+      end
+
+      it "returns false for a response with a different code" do
+        response = create_response(:status => other_code)
+        expect( matcher.matches?(response) ).to be(false)
+      end
+    end
+
+    it "describes #{description}" do
+      expect(matcher.description).to eq(description)
+    end
+
+    it "has a failure message reporting the expected and actual status codes" do
+      response = create_response(:status => other_code)
+      expect{ matcher.matches? response }.
+        to change(matcher, :failure_message).
+        to(failure_message)
+    end
+
+    it "has a negated failure message reporting the expected and actual status codes" do
+      response = create_response(:status => expected_code)
+      expect{ matcher.matches? response }.
+        to change(matcher, :failure_message_when_negated).
+        to(negated_message)
+    end
+  end
+
```

Comment:
The diff from here out looks terrible. All that happened is:

- the repeated code in this shared example was replaced with invocation of this shared example
- new tests using the shared example were added for :not_found, :server_error, and :successful